test/src/util/log_monitor.rs

```diff
@@ -0,0 +1,55 @@
+use crate::node::Node;
+use crate::utils::{node_log, sleep, tweaked_duration};
+
+use std::fs::File;
+use std::io::{BufRead, BufReader, Seek, SeekFrom};
+use std::time::Instant;
+
+use crossbeam_channel::{bounded, Receiver};
+
+pub fn monitor_log_until_expected_show(
+    node: &Node,
+    seek_from: u64,
+    timeout: u64,
+    expected: &str,
+) -> Receiver<bool> {
+    let predicate = |file_reader: BufReader<&File>| {
+        file_reader
+            .lines()
+            .any(|line| line.unwrap().contains(expected))
+    };
+    monitor_log_until_bool(node, seek_from, timeout, predicate)
+}
+
+fn monitor_log_until_bool<P>(
+    node: &Node,
+    mut seek_from: u64,
+    timeout: u64,
+    predicate: P,
+) -> Receiver<bool>
+where
+    P: Fn(BufReader<&File>) -> bool,
+{
+    let (log_sender, log_notifier) = bounded(1);
+    let timeout = tweaked_duration(timeout);
+    let start = Instant::now();
+    let filename = node_log(node.working_dir());
+    let mut file = File::open(&filename).unwrap();
+    loop {
+        let file_size = node.log_size();
+        if seek_from != file_size {
+            file.seek(SeekFrom::Start(seek_from)).unwrap();
+            let file_reader = BufReader::new(&file);
+            if predicate(file_reader) {
+                let _ = log_sender.send(true);
+                return log_notifier;
```

Comment:
Why return a `log_notifier` but not just return `true`?

Reply:
Good question. I thought about this and I think we don't even need the channel in this condition, so I removed crossbeam and return an `Option<String>`. And return the `String` because sometime we may want to extract something from log.

Reply:
Nice. However, after your changing to return `Option<S>`, the function name `*_bool`(`monitor_log_until_bool`) becomes weird.

Reply:
yeah, updated, pls review again, thx